File: 0279-perfect-squares/0279-perfect-squares.cpp
class Solution {
public:
    vector<vector<int>> memo;
    int numSquares(int n) {
        vector<int> nums;
        for (int i = 1; i * i <= n; i++) {
            nums.push_back(i * i);
        }
        memo = vector<vector<int>>(nums.size(), vector<int>(n + 1));
        return solve(n, nums, 0);
    }

    int solve(int n, vector<int>& nums, int index) {
        if (n == 0) {
            return 0;
        }
        if (index >= nums.size() || n < 0) {
            return 100000;
        }
        if (memo[index][n]) {
            return memo[index][n];
        }

        int take = 1 + solve(n - nums[index], nums, index);
        int skip = solve(n, nums, index + 1);

        return memo[index][n] = min(take, skip);
    }
};

File: 0279-perfect-squares/0279-perfect-squares.py
class Solution:
    def numSquares(self, n: int) -> int:
        arr = []
        for i in range(1, int(math.sqrt(n)) + 1):
            arr.append(i * i)
        
        arr = arr[::-1]

        dp = [[0 for _ in range(n + 1)] for _ in range(len(arr) + 1)]

        def solve(n, idx):
            if idx == len(arr) or n < 0:
                return inf

            if n - arr[idx] == 0:
                return 1

            
            if dp[idx][n]:
                return dp[idx][n]

            take = skip = 0
            take = 1 + solve(n - arr[idx], idx)
            skip = solve(n, idx + 1)

            dp[idx][n] = min(take, skip)
            return min(take, skip)

        return solve(n, 0)
